### ISSUE TITLE
nn modules inlining effect freezing inductor path

### ISSUE THREAD

**Issue by laithsakka** [pytorch/pytorch#126093]
There are several failures with AssertionError: Scalars are not equal!, not sure if they are all the same issue will figure out once i fix this one. 

**Note: the comments in the PR are for me, they are not questions for anyone ! unless someone is tagged**

cc @ezyang @msaroufim @bdhirsh @anijain2305 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire

**Comment by laithsakka** [comment]
Looked at the test,
```
 TORCHDYNAMO_INLINE_INBUILT_NN_MODULES=1 python test/inductor/test_cpu_cpp_wrapper.py -k test_qlinear_relu_cpu_dynamic_shapes_cpp_wrapper
 ```
 
fill output log is
```
base) [lsakka@devvm4561.ash0 /data/users/lsakka/pytorch/pytorch (fix_msg)]$ TORCHDYNAMO_INLINE_INBUILT_NN_MODULES=1 TORCH_LOGS="graph" python test/inductor/test_cpu_cpp_wrapper.py -k test_qlinear_relu_cpu_dynamic_shapes_cpp_wrapper
W0513 15:15:45.763000 139844412389184 torch/_export/__init__.py:94] +============================+
W0513 15:15:45.764000 139844412389184 torch/_export/__init__.py:95] |     !!!   WARNING   !!!    |
W0513 15:15:45.764000 139844412389184 torch/_export/__init__.py:96] +============================+
W0513 15:15:45.764000 139844412389184 torch/_export/__init__.py:97] capture_pre_autograd_graph() is deprecated and doesn't provide any function guarantee moving forward.
W0513 15:15:45.764000 139844412389184 torch/_export/__init__.py:98] Please switch to use torch.export instead.
W0513 15:15:47.723000 139844412389184 torch/_dynamo/exc.py:205] [7/0] Backend compiler failed with a fake tensor exception at 
W0513 15:15:47.723000 139844412389184 torch/_dynamo/exc.py:205] [7/0]   File "<eval_with_key>.74", line 25, in forward
W0513 15:15:47.723000 139844412389184 torch/_dynamo/exc.py:205] [7/0]     return pytree.tree_unflatten([relu_1], self._out_spec)
W0513 15:15:47.723000 139844412389184 torch/_dynamo/exc.py:205] [7/0] Adding a graph break.
W0513 15:15:48.121000 139844412389184 torch/_dynamo/exc.py:205] [7/0_1] Backend compiler failed with a fake tensor exception at 
W0513 15:15:48.121000 139844412389184 torch/_dynamo/exc.py:205] [7/0_1]   File "<eval_with_key>.74", line 25, in forward
W0513 15:15:48.121000 139844412389184 torch/_dynamo/exc.py:205] [7/0_1]     return pytree.tree_unflatten([relu_1], self._out_spec)
W0513 15:15:48.121000 139844412389184 torch/_dynamo/exc.py:205] [7/0_1] Adding a graph break.
stats [('calls_captured', 34), ('unique_graphs', 7)]
inline_call []
frames [('total', 1)]
inductor [('qlinear_weight_prepack_matcher_nodes', 16), ('pattern_matcher_nodes', 16), ('qlinear_weight_prepack_matcher_count', 4), ('pattern_matcher_count', 4), ('fxgraph_cache_bypass', 2)]
aot_autograd [('total', 2), ('not_ok', 2)]
unimplemented [('Backend compiler failed with a fake tensor exception at \n  File "<eval_with_key>.74", line 25, in forward\n    return pytree.tree_unflatten([relu_1], self._out_spec)\nAdding a graph break.', 2)]
F
======================================================================
FAIL: test_qlinear_relu_cpu_dynamic_shapes_cpp_wrapper (__main__.DynamicShapesCppWrapperCpuTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/lsakka/pytorch/pytorch/torch/testing/_internal/common_utils.py", line 2756, in wrapper
    method(*args, **kwargs)
  File "/data/users/lsakka/pytorch/pytorch/test/inductor/test_torchinductor.py", line 10008, in new_test
    return value(self)
  File "/data/users/lsakka/pytorch/pytorch/torch/_dynamo/testing.py", line 309, in _fn
    return fn(*args, **kwargs)
  File "/data/users/lsakka/miniconda3/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/data/users/lsakka/pytorch/pytorch/test/inductor/test_cpu_cpp_wrapper.py", line 137, in fn
    _, code = test_torchinductor.run_and_get_cpp_code(
  File "/data/users/lsakka/pytorch/pytorch/test/inductor/test_torchinductor.py", line 357, in run_and_get_cpp_code
    result = fn(*args, **kwargs)
  File "/data/users/lsakka/pytorch/pytorch/torch/testing/_internal/common_quantization.py", line 420, in wrapper
    fn(*args, **kwargs)
  File "/data/users/lsakka/pytorch/pytorch/torch/testing/_internal/common_quantization.py", line 372, in wrapper
    fn(*args, **kwargs)
  File "/data/users/lsakka/pytorch/pytorch/torch/testing/_internal/common_utils.py", line 1560, in wrapper
    return fn(*args, **kwargs)
  File "/data/users/lsakka/pytorch/pytorch/test/inductor/test_mkldnn_pattern_matcher.py", line 1534, in test_qlinear_relu_cpu
    self._qlinear_unary_cpu_test_helper((torch.randn((2, 4)),))
  File "/data/users/lsakka/pytorch/pytorch/test/inductor/test_mkldnn_pattern_matcher.py", line 1519, in _qlinear_unary_cpu_test_helper
    self._test_common(
  File "/data/users/lsakka/pytorch/pytorch/test/inductor/test_mkldnn_pattern_matcher.py", line 177, in _test_common
    matcher_check_fn()
  File "/data/users/lsakka/pytorch/pytorch/test/inductor/test_mkldnn_pattern_matcher.py", line 1513, in matcher_check_fn
    self.assertEqual(
  File "/data/users/lsakka/pytorch/pytorch/torch/testing/_internal/common_utils.py", line 3639, in assertEqual
    raise error_metas.pop()[0].to_error(
AssertionError: Scalars are not equal!

Expected 2 but got 4.
Absolute difference: 2
Relative difference: 1.0

To execute this test, run the following from the base repo dir:
     python test/inductor/test_cpu_cpp_wrapper.py -k DynamicShapesCppWrapperCpuTests.test_qlinear_relu_cpu_dynamic_shapes_cpp_wrapper

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

----------------------------------------------------------------------
Ran 1 test in 2.373s
```

**Comment by laithsakka** [comment]
The reason for the graph break is the following:
UnsupportedOperatorException(func=<OpOverload(op='onednn.qlinear_prepack', overload='default')>)

**Comment by laithsakka** [comment]
Ok I looked at this, and here is what is happening.
1) when  config.freezing is enabled, some pass runs in inductor freeze() that  transforms the compiled dynamo graph and adds qlinear_prepack (during qlinear_weight_prepack pass)

2)  Then after that we call compile_fx_inner after qlinear_prepack is introduced.
3) inside fx_codegen_and_compile, we call fake_tensor_prop with the following assumption:
```
    # It is safe to run FakeTensorProp under no_grad because by the time
    # we're in inductor, we assume that AOTAutograd has already "taken care"
    # of autograd, so there should be no more autograd-related API's in the
    # graph.
    with torch.no_grad():
        fake_mode = fake_tensor_prop(gm, example_inputs)
 ```
4) when running qlinear_prepack we get UnsupportedOperatorException, hence the compilation fails
5 We propagated the error back to dynamo and adds a graph break!


Now the question is why is it that when nn module inlining is disabled we do not run into this?
to answer that I printed the graph for that goes through fake_tesnor_prop before with nn-module inlining and without it:
without
```
class <lambda>(torch.nn.Module):
    def forward(self, arg8_1: "Sym(s0)", arg9_1: "f32[s0, 4]"):
        # No stacktrace found for following nodes
        _frozen_param0: "f32[4]" = self._frozen_param0
        _frozen_param1: "f32[4]" = self._frozen_param1
        _frozen_param2: "f32[4]" = self._frozen_param2
        _frozen_param3: "i64[4]" = self._frozen_param3
        _frozen_param5: "f32[4]" = self._frozen_param5
        _frozen_param6: "i64[4]" = self._frozen_param6
        
        # File: <eval_with_key>.74:7 in forward, code: quantize_per_tensor_default = torch.ops.quantized_decomposed.quantize_per_tensor.default(arg0_1, 0.006681466940790415, 151, 0, 255, torch.uint8);  arg0_1 = None
        quantize_per_tensor: "u8[s0, 4]" = torch.ops.quantized_decomposed.quantize_per_tensor.default(arg9_1, 0.006681466940790415, 151, 0, 255, torch.uint8);  arg9_1 = None
        
        # No stacktrace found for following nodes
        _frozen_param8 = self._frozen_param8
        
        # File: /data/users/lsakka/pytorch/pytorch/test/inductor/test_mkldnn_pattern_matcher.py:1504 in forward, code: tmp = self.unary_fn(self.linear(x))
        qlinear_pointwise_default_1: "f32[s0, 4]" = torch.ops.onednn.qlinear_pointwise.default(quantize_per_tensor, 0.006681466940790415, 151, _frozen_param8, _frozen_param2, _frozen_param3, _frozen_param0, 1.0, 0, torch.float32, 'none', [], '');  quantize_per_tensor = _frozen_param8 = _frozen_param2 = _frozen_param3 = _frozen_param0 = None
        relu: "f32[s0, 4]" = torch.ops.aten.relu.default(qlinear_pointwise_default_1);  qlinear_pointwise_default_1 = None
        
        # File: <eval_with_key>.74:16 in forward, code: quantize_per_tensor_default_1 = torch.ops.quantized_decomposed.quantize_per_tensor.default(relu, 0.0025742885190993547, 0, 0, 255, torch.uint8);  relu = None
        quantize_per_tensor_1: "u8[s0, 4]" = torch.ops.quantized_decomposed.quantize_per_tensor.default(relu, 0.0025742885190993547, 0, 0, 255, torch.uint8);  relu = None
        
        # No stacktrace found for following nodes
        _frozen_param9 = self._frozen_param9
        
        # File: /data/users/lsakka/pytorch/pytorch/test/inductor/test_mkldnn_pattern_matcher.py:1505 in forward, code: return self.unary_fn2(self.linear2(tmp))
        qlinear_pointwise_default: "f32[s0, 4]" = torch.ops.onednn.qlinear_pointwise.default(quantize_per_tensor_1, 0.0025742885190993547, 0, _frozen_param9, _frozen_param5, _frozen_param6, _frozen_param1, 1.0, 0, torch.float32, 'none', [], '');  quantize_per_tensor_1 = _frozen_param9 = _frozen_param5 = _frozen_param6 = _frozen_param1 = None
        relu_1: "f32[s0, 4]" = torch.ops.aten.relu.default(qlinear_pointwise_default);  qlinear_pointwise_default = None
        return (relu_1,)
   ```
   
   with 
   ```
   class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: "Sym(s0)", arg1_1: "Sym(s2)", arg2_1: "Sym(s3)", arg3_1: "f32[s3, 4]", arg4_1: "f32[4]", arg5_1: "i64[4]", arg6_1: "i8[4, 4]", arg7_1: "f32[4]", arg8_1: "f32[4]", arg9_1: "i64[4]", arg10_1: "i8[4, 4]", arg11_1: "f32[4]"):
        # File: <eval_with_key>.74:7 in forward, code: quantize_per_tensor_default = torch.ops.quantized_decomposed.quantize_per_tensor.default(arg0_1, 0.006681466940790415, 151, 0, 255, torch.uint8);  arg0_1 = None
        quantize_per_tensor: "u8[s3, 4]" = torch.ops.quantized_decomposed.quantize_per_tensor.default(arg3_1, 0.006681466940790415, 151, 0, 255, torch.uint8);  arg3_1 = None
        
        # No stacktrace found for following nodes
        qlinear_prepack_1 = torch.ops.onednn.qlinear_prepack(arg6_1, None);  arg6_1 = None
        
        # File: /data/users/lsakka/pytorch/pytorch/test/inductor/test_mkldnn_pattern_matcher.py:1504 in forward, code: tmp = self.unary_fn(self.linear(x))
        qlinear_pointwise_default_1: "f32[s3, 4]" = torch.ops.onednn.qlinear_pointwise.default(quantize_per_tensor, 0.006681466940790415, 151, qlinear_prepack_1, arg4_1, arg5_1, arg7_1, 1.0, 0, torch.float32, 'none', [], '');  quantize_per_tensor = qlinear_prepack_1 = arg4_1 = arg5_1 = arg7_1 = None
        relu: "f32[s3, 4]" = torch.ops.aten.relu.default(qlinear_pointwise_default_1);  qlinear_pointwise_default_1 = None
        
        # File: <eval_with_key>.74:16 in forward, code: quantize_per_tensor_default_1 = torch.ops.quantized_decomposed.quantize_per_tensor.default(relu, 0.0025742885190993547, 0, 0, 255, torch.uint8);  relu = None
        quantize_per_tensor_1: "u8[s3, 4]" = torch.ops.quantized_decomposed.quantize_per_tensor.default(relu, 0.0025742885190993547, 0, 0, 255, torch.uint8);  relu = None
        
        # No stacktrace found for following nodes
        qlinear_prepack = torch.ops.onednn.qlinear_prepack(arg10_1, None);  arg10_1 = None
        
        # File: /data/users/lsakka/pytorch/pytorch/test/inductor/test_mkldnn_pattern_matcher.py:1505 in forward, code: return self.unary_fn2(self.linear2(tmp))
        qlinear_pointwise_default: "f32[s3, 4]" = torch.ops.onednn.qlinear_pointwise.default(quantize_per_tensor_1, 0.0025742885190993547, 0, qlinear_prepack, arg8_1, arg9_1, arg11_1, 1.0, 0, torch.float32, 'none', [], '');  quantize_per_tensor_1 = qlinear_prepack = arg8_1 = arg9_1 = arg11_1 = None
        relu_1: "f32[s3, 4]" = torch.ops.aten.relu.default(qlinear_pointwise_default);  qlinear_pointwise_default = None
        return (relu_1,)
        
        ```

**Comment by laithsakka** [comment]
the difference i see its that its effecting frozen params.
     qlinear_prepack_1 = torch.ops.onednn.qlinear_prepack(arg6_1, None);  arg6_1 = None

vs
        _frozen_param8 = self._frozen_param8
       
Next I am looking at that

**Comment by laithsakka** [comment]
ok the issue I am seeing is that specialization is not being triggered when nn module inlining is enabled (frozen arguments are dynamic when nn-module inlining enabled, but not dynamic otherwise). 

(before freeze code is) 
```
entry to freeze
class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: "f32[4]", arg1_1: "f32[4]", arg2_1: "f32[4]", arg3_1: "i64[4]", arg4_1: "i8[4, 4]", arg5_1: "f32[4]", arg6_1: "i64[4]", arg7_1: "i8[4, 4]", arg8_1: "Sym(s0)", arg9_1: "f32[s0, 4]"):
        # File: <eval_with_key>.74:7 in forward, code: quantize_per_tensor_default = torch.ops.quantized_decomposed.quantize_per_tensor.default(arg0_1, 0.006681466940790415, 151, 0, 255, torch.uint8);  arg0_1 = None
        quantize_per_tensor: "u8[s0, 4]" = torch.ops.quantized_decomposed.quantize_per_tensor.default(arg9_1, 0.006681466940790415, 151, 0, 255, torch.uint8);  arg9_1 = None
        
        # File: /data/users/lsakka/pytorch/pytorch/test/inductor/test_mkldnn_pattern_matcher.py:1504 in forward, code: tmp = self.unary_fn(self.linear(x))
        dequantize_per_tensor: "f32[s0, 4]" = torch.ops.quantized_decomposed.dequantize_per_tensor.default(quantize_per_tensor, 0.006681466940790415, 151, 0, 255, torch.uint8);  quantize_per_tensor = None
        dequantize_per_channel: "f32[4, 4]" = torch.ops.quantized_decomposed.dequantize_per_channel.default(arg4_1, arg2_1, arg3_1, 0, -128, 127, torch.int8);  arg4_1 = arg2_1 = arg3_1 = None
        permute: "f32[4, 4]" = torch.ops.aten.permute.default(dequantize_per_channel, [1, 0]);  dequantize_per_channel = None
        addmm: "f32[s0, 4]" = torch.ops.aten.addmm.default(arg0_1, dequantize_per_tensor, permute);  arg0_1 = dequantize_per_tensor = permute = None
        relu: "f32[s0, 4]" = torch.ops.aten.relu.default(addmm);  addmm = None
        
        # File: <eval_with_key>.74:16 in forward, code: quantize_per_tensor_default_1 = torch.ops.quantized_decomposed.quantize_per_tensor.default(relu, 0.0025742885190993547, 0, 0, 255, torch.uint8);  relu = None
        quantize_per_tensor_1: "u8[s0, 4]" = torch.ops.quantized_decomposed.quantize_per_tensor.default(relu, 0.0025742885190993547, 0, 0, 255, torch.uint8);  relu = None
        
        # File: /data/users/lsakka/pytorch/pytorch/test/inductor/test_mkldnn_pattern_matcher.py:1505 in forward, code: return self.unary_fn2(self.linear2(tmp))
        dequantize_per_tensor_1: "f32[s0, 4]" = torch.ops.quantized_decomposed.dequantize_per_tensor.default(quantize_per_tensor_1, 0.0025742885190993547, 0, 0, 255, torch.uint8);  quantize_per_tensor_1 = None
        dequantize_per_channel_1: "f32[4, 4]" = torch.ops.quantized_decomposed.dequantize_per_channel.default(arg7_1, arg5_1, arg6_1, 0, -128, 127, torch.int8);  arg7_1 = arg5_1 = arg6_1 = None
        permute_1: "f32[4, 4]" = torch.ops.aten.permute.default(dequantize_per_channel_1, [1, 0]);  dequantize_per_channel_1 = None
        addmm_1: "f32[s0, 4]" = torch.ops.aten.addmm.default(arg1_1, dequantize_per_tensor_1, permute_1);  arg1_1 = dequantize_per_tensor_1 = permute_1 = None
        relu_1: "f32[s0, 4]" = torch.ops.aten.relu.default(addmm_1);  addmm_1 = None
        return (relu_1,)
```

```
class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: "Sym(s0)", arg1_1: "Sym(s2)", arg2_1: "Sym(s3)", arg3_1: "f32[s3, 4]", arg4_1: "f32[4]", arg5_1: "i64[4]", arg6_1: "i8[4, 4]", arg7_1: "f32[4]", arg8_1: "f32[4]", arg9_1: "i64[4]", arg10_1: "i8[4, 4]", arg11_1: "f32[4]"):
        # File: <eval_with_key>.74:7 in forward, code: quantize_per_tensor_default = torch.ops.quantized_decomposed.quantize_per_tensor.default(arg0_1, 0.006681466940790415, 151, 0, 255, torch.uint8);  arg0_1 = None
        quantize_per_tensor: "u8[s3, 4]" = torch.ops.quantized_decomposed.quantize_per_tensor.default(arg3_1, 0.006681466940790415, 151, 0, 255, torch.uint8);  arg3_1 = None
        
        # File: /data/users/lsakka/pytorch/pytorch/test/inductor/test_mkldnn_pattern_matcher.py:1504 in forward, code: tmp = self.unary_fn(self.linear(x))
        dequantize_per_tensor: "f32[s3, 4]" = torch.ops.quantized_decomposed.dequantize_per_tensor.default(quantize_per_tensor, 0.006681466940790415, 151, 0, 255, torch.uint8);  quantize_per_tensor = None
        dequantize_per_channel: "f32[4, 4]" = torch.ops.quantized_decomposed.dequantize_per_channel.default(arg6_1, arg4_1, arg5_1, 0, -128, 127, torch.int8);  arg6_1 = arg4_1 = arg5_1 = None
        permute: "f32[4, 4]" = torch.ops.aten.permute.default(dequantize_per_channel, [1, 0]);  dequantize_per_channel = None
        addmm: "f32[s3, 4]" = torch.ops.aten.addmm.default(arg7_1, dequantize_per_tensor, permute);  arg7_1 = dequantize_per_tensor = permute = None
        relu: "f32[s3, 4]" = torch.ops.aten.relu.default(addmm);  addmm = None
        
        # File: <eval_with_key>.74:16 in forward, code: quantize_per_tensor_default_1 = torch.ops.quantized_decomposed.quantize_per_tensor.default(relu, 0.0025742885190993547, 0, 0, 255, torch.uint8);  relu = None
        quantize_per_tensor_1: "u8[s3, 4]" = torch.ops.quantized_decomposed.quantize_per_tensor.default(relu, 0.0025742885190993547, 0, 0, 255, torch.uint8);  relu = None
        
        # File: /data/users/lsakka/pytorch/pytorch/test/inductor/test_mkldnn_pattern_matcher.py:1505 in forward, code: return self.unary_fn2(self.linear2(tmp))
        dequantize_per_tensor_1: "f32[s3, 4]" = torch.ops.quantized_decomposed.dequantize_per_tensor.default(quantize_per_tensor_1, 0.0025742885190993547, 0, 0, 255, torch.uint8);  quantize_per_tensor_1 = None
        dequantize_per_channel_1: "f32[4, 4]" = torch.ops.quantized_decomposed.dequantize_per_channel.default(arg10_1, arg8_1, arg9_1, 0, -128, 127, torch.int8);  arg10_1 = arg8_1 = arg9_1 = None
        permute_1: "f32[4, 4]" = torch.ops.aten.permute.default(dequantize_per_channel_1, [1, 0]);  dequantize_per_channel_1 = None
        addmm_1: "f32[s3, 4]" = torch.ops.aten.addmm.default(arg11_1, dequantize_per_tensor_1, permute_1);  arg11_1 = dequantize_per_tensor_1 = permute_1 = None
        relu_1: "f32[s3, 4]" = torch.ops.aten.relu.default(addmm_1);  addmm_1 = None
        return (relu_1,)
        ```
 Next looking into why the inputs are dynamic only when inlining is enabled.

**Comment by laithsakka** [comment]
I digged into this and found the issue:
in replace_params_with_constants() we actually replace the input params with the their constant values. 
```

def replace_params_with_constants(
    gm: torch.fx.GraphModule,
    flat_params: list[Any],
    fw_metadata: torch._functorch.aot_autograd.ViewAndMutationMeta,
) -> List[int]:
    """
    Replaces the parameters of a PyTorch GraphModule with constants wherever possible.
    Returns a list of indices representing the input parameters that were not converted to constants.
    """
    params = gm.graph.find_nodes(op="placeholder")
    fake_inp_nodes = params[: len(params)]
    preserved_arg_indices = []
    aliased_input_args = [
        out_info.base_idx
        for out_info in fw_metadata.output_info
        if out_info.base_idx is not None
    ]

    # TODO (tmanlaibaatar) figure out why this is different
    # from mutated_inp_runtime_indices
    mutated_inps = [
        i
        for i, m in enumerate(fw_metadata.input_info)
        if m.mutation_type
        in (MutationType.MUTATED_IN_GRAPH, MutationType.MUTATED_OUT_GRAPH)
    ]

    for i, (real_input, node) in enumerate(zip(flat_params, fake_inp_nodes)):
        if i in mutated_inps or i in aliased_input_args:
            preserved_arg_indices.append(i)
            continue
        replace_node_with_constant(gm, node, real_input)
    # add on non param inputs
    preserved_arg_indices.extend(range(len(flat_params), len(params)))
    # is this necessary ?
    gm.recompile()
    return preserved_arg_indices
    ```
but when nn module inlining is enabled, flat_params is actually an empty list! why? because we read those params here:
```
def aot_module_simplified(
    mod: nn.Module,
    args,
    fw_compiler: Callable,
    bw_compiler: Optional[Callable] = None,
    partition_fn: Callable = default_partition,
    decompositions: Optional[Dict] = None,
    keep_inference_input_mutations=False,
    inference_compiler: Optional[Callable] = None,
) -> nn.Module:
    """
    This is the simplified or low overhead version of aot_module. For frontends
    like TorchDynamo, the input functions/modules to AOT are static and have
    unpacked inputs/outputs. This gives us an opportunity to remove the
        (1) pytree overhead to parse inputs/outputs,
        (2) AOT Autograd cache,
        (3) Reading of params/buffers in every forward call

    :func:`aot_module_simplified` removes these overheads.
    """
    params = {
        **dict(mod.named_parameters(remove_duplicate=False)),
        **dict(mod.named_buffers(remove_duplicate=False)),
    }
    params_flat, params_spec = pytree.tree_flatten(params)
    params_flat = list(params_flat)
    params_len = len(params_flat)
```
and in this case the params is empty list because those params with inlining become object variables(I think)
Those are read here:


```

def aot_module_simplified(
    mod: nn.Module,
    args,
    fw_compiler: Callable,
    bw_compiler: Optional[Callable] = None,
    partition_fn: Callable = default_partition,
    decompositions: Optional[Dict] = None,
    keep_inference_input_mutations=False,
    inference_compiler: Optional[Callable] = None,
) -> nn.Module:
    """
    This is the simplified or low overhead version of aot_module. For frontends
    like TorchDynamo, the input functions/modules to AOT are static and have
    unpacked inputs/outputs. This gives us an opportunity to remove the
        (1) pytree overhead to parse inputs/outputs,
        (2) AOT Autograd cache,
        (3) Reading of params/buffers in every forward call

    :func:`aot_module_simplified` removes these overheads.
    """
    params = {
        **dict(mod.named_parameters(remove_duplicate=False)),
        **dict(mod.named_buffers(remove_duplicate=False)),
    }
    params_flat, params_spec = pytree.tree_flatten(params)
    params_flat = list(params_flat)
    params_len = len(params_flat)

    functional_call = create_functional_call(mod, params_spec, params_len)

    if bw_compiler is None:
        bw_compiler = fw_compiler
    if inference_compiler is None:
        inference_compiler = fw_compiler

    seen_sources = set()
  ```
  Note that inputs to forward are all prefixed with self, and mod.named_parameters , mod.named_buffersi aempty
  ```
  '\n\n\ndef forward(self, L_self_in_spec_num_nodes, L_self_in_spec_children_specs_0_num_nodes, s3, L_x_, L_self_linear_scale_0, L_self_linear_zero_point_0, L_self_frozen_param0, L_self_linear_bias, L_self_linear2_scale_0, L_self_linear2_zero_point_0, L_self_frozen_param1, L_self_linear2_bias):\n    l_self_in_spec_num_nodes = L_self_in_spec_num_nodes\n    l_self_in_spec_children_specs_0_num_nodes = L_self_in_spec_children_specs_0_num_nodes\n    l_x_ = L_x_\n    l_self_linear_scale_0 = L_self_linear_scale_0\n    l_self_linear_zero_point_0 = L_self_linear_zero_point_0\n    l_self_frozen_param0 = L_self_frozen_param0\n    l_self_linear_bias = L_self_linear_bias\n    l_self_linear2_scale_0 = L_self_linear2_scale_0\n    l_self_linear2_zero_point_0 = L_self_linear2_zero_point_0\n    l_self_frozen_param1 = L_self_frozen_param1\n    l_self_linear2_bias = L_self_linear2_bias\n    eq = l_self_in_spec_num_nodes == 1;  l_self_in_spec_num_nodes = None\n    eq_1 = l_self_in_spec_children_specs_0_num_nodes == 1;  l_self_in_spec_children_specs_0_num_nodes = None\n    quantize_per_tensor_default = torch.ops.quantized_decomposed.quantize_per_tensor.default(l_x_, 0.006681466940790415, 151, 0, 255, torch.uint8);  l_x_ = None\n    dequantize_per_tensor_default = torch.ops.quantized_decomposed.dequantize_per_tensor.default(quantize_per_tensor_default, 0.006681466940790415, 151, 0, 255, torch.uint8);  quantize_per_tensor_default = None\n    dequantize_per_channel_default = torch.ops.quantized_decomposed.dequantize_per_channel.default(l_self_frozen_param0, l_self_linear_scale_0, l_self_linear_zero_point_0, 0, -128, 127, torch.int8);  l_self_frozen_param0 = l_self_linear_scale_0 = l_self_linear_zero_point_0 = None\n    linear_default = torch.ops.aten.linear.default(dequantize_per_tensor_default, dequantize_per_channel_default, l_self_linear_bias);  dequantize_per_tensor_default = dequantize_per_channel_default = l_self_linear_bias = None\n    relu_default = torch.ops.aten.relu.default(linear_default);  linear_default = None\n    quantize_per_tensor_default_1 = torch.ops.quantized_decomposed.quantize_per_tensor.default(relu_default, 0.0025742885190993547, 0, 0, 255, torch.uint8);  relu_default = None\n    dequantize_per_tensor_default_1 = torch.ops.quantized_decomposed.dequantize_per_tensor.default(quantize_per_tensor_default_1, 0.0025742885190993547, 0, 0, 255, torch.uint8);  quantize_per_tensor_default_1 = None\n    dequantize_per_channel_default_1 = torch.ops.quantized_decomposed.dequantize_per_channel.default(l_self_frozen_param1, l_self_linear2_scale_0, l_self_linear2_zero_point_0, 0, -128, 127, torch.int8);  l_self_frozen_param1 = l_self_linear2_scale_0 = l_self_linear2_zero_point_0 = None\n    linear_default_1 = torch.ops.aten.linear.default(dequantize_per_tensor_default_1, dequantize_per_channel_default_1, l_self_linear2_bias);  dequantize_per_tensor_default_1 = dequantize_per_channel_default_1 = l_self_linear2_bias = None\n    relu_default_1 = torch.ops.aten.relu.default(linear_default_1);  linear_default_1 = None\n    return (relu_default_1,)\n    '

  ```

**Comment by laithsakka** [comment]
I chatted with @mlazos  about this, he is working on a change to identify params and buffers in autograd, after that this should be easy to fix

**Comment by eellison** [comment]
I don't know that we should support the combination of the two. freezing makes parameters constants in the graph. dont know if theres an easy way of sharing with not inlining nn modules.

**Comment by mlazos** [comment]
FYI I discussed this with @eellison and @anijain2305 and the POR is to guard on parameter locations in these cases. 

The problem is as follows: 
For each instance of a builtin NNModule we will get a single graph with module inlining enabled with params as inputs. As a result we will no longer constant fold them. There are 3 options here: 1) don't constant fold (bad) 2) recompile or 3) implement a dispatching layer where we have multiple graphs with the different sets of params constant folded

Since freezing is rare we opted to go with 2, since 3 has many more unknowns and 2 will not affect performance, since it is the status quo.
